src/dataframe/dataframe.jl

```diff
@@ -901,7 +903,7 @@ end
 function hcat!(df1::DataFrame, df2::AbstractDataFrame; makeunique::Bool=false)
     u = add_names(index(df1), index(df2), makeunique=makeunique)
     for i in 1:length(u)
-        df1[u[i]] = df2[i]
+        df1[u[i]] = copy(df2[i])
```

Comment:
This appears to be the most problematic change, since we don't provide an alternative syntax to avoid making copies. Maybe we could just add a keyword argument?

Reply:
`hcat!` is not exported. So the key discussion is about `hcat`. Now it always copies everything it gets and this is consistent with the contract.

If someone really wants to create a new `DataFrame` from two or more existing `DataFrames` without copying then `eachcol(df, true)` should be iterated over (this is clumsy, but I guess we agree that creating a data frame reusing columns from another data frame is error prone and should not be easy to do).

Reply:
Yes I was thinking of `hcat`. Using `eachcol(df, true)` sounds really impractical (and hard to discover). I agree that sharing columns should not be too easy, or at least that the most obvious solution should make copies. But I also think we should provide simple and discoverable ways of avoiding copies for all operations, just like we do with `view`. I think that's consistent with the spirit of `@inbounds`: it's easy to shoot yourself in the foot if you want, but it's not the default.

Overall I think we should have a general plan regarding copies. We're not too far from that I guess as your text in this PR shows. Maybe we could use the same keyword argument in `hcat` and in `stack`/`melt` to avoid making copies, which will allow removing `stackdf` and `meltdf` (https://github.com/JuliaData/DataFrames.jl/issues/1736)?

Reply:
Maybe `reusecolumns` kwarg name?

Reply:
Or `copycolumns`, or just `copy`? We could even add it to the constructor at some point (with a default to `false`).

Reply:
Good point with the constructor. I would call it `copycolumns` then as:
1. it is long - so it will not be used lightly 😄 
2. `copy` is a conflicting name

Reply:
Well there's no real conflict with keyword arguments, but why not.

Reply:
Well - actually we use `copy` function inside this function 😃.